### PR TITLE
fix SingleChildScrollView clip bug

### LINKED ISSUE #61434: TextField’s border is cut if placed inside a SingleChildScrollView
When a `TextField` is the bottommost widget in a `SingleChildScrollView`, its bottom border appears to be thinner than usual. Probably a rendering issue.

An unwrapped `TextField` looks like this:

Screenshot 2020-07-14 at 12 11 11

A `TextField` inside a `SingleChildScrollView` looks like this:

Screenshot 2020-07-14 at 12 11 20

If the border width is less than `1.0`, then it becomes almost invisible inside the scroll view.

Here’s a sample code to reproduce the issue:
```
import 'package:flutter/material.dart';

void main() {
  runApp(SampleWidget());
}

class SampleWidget extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      builder: (_, __) => Scaffold(
        body: Center(
          child: Padding(
            padding: const EdgeInsets.all(25.0),
            child: SingleChildScrollView(
              child: TextField(),
            ),
          ),
        ),
      ),
    );
  }
}
```

<details><summary><code>flutter doctor -v</code></summary>
<pre>
[✓] Flutter (Channel stable, v1.17.5, on Mac OS X 10.15.5 19F101, locale en-GB)
    • Flutter version 1.17.5 at /Users/denis/Projects/vendor/flutter
    • Framework revision 8af6b2f038 (2 weeks ago), 2020-06-30 12:53:55 -0700
    • Engine revision ee76268252
    • Dart version 2.8.4

 
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at /Users/denis/Library/Android/sdk
    • Platform android-30, build-tools 29.0.2
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.5)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.5, Build version 11E608c
    • CocoaPods version 1.8.4

[✓] Android Studio (version 3.5)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 41.1.2
    • Dart plugin version 191.8593
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)

[!] IntelliJ IDEA Ultimate Edition (version 2020.1.3)
    • IntelliJ at /Applications/IntelliJ IDEA.app
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
    • For information about installing plugins, see
      https://flutter.dev/intellij-setup/#installing-the-plugins

[✓] Connected device (2 available)
    • Android SDK built for x86 • emulator-5554                        • android-x86 • Android 10 (API 29) (emulator)
    • iPhone 11 Pro Max         • 6FCA75B1-A34C-45EC-BF25-BA36CC5DA55D • ios         • com.apple.CoreSimulator.SimRuntime.iOS-13-5 (simulator)

! Doctor found issues in 1 category.
</pre>
</details>

### PR DESCRIPTION
**Pull request by xu-baolin**
## Description

```
  bool _shouldClipAtPaintOffset(Offset paintOffset) {
    assert(child != null);
    return paintOffset < Offset.zero || !(Offset.zero & size).contains((paintOffset & child.size).bottomRight);
  }
```

This function has two bugs:
1. `paintOffset < Offset.zero` The result is not expected
2. `Rect.contains` exclude their bottom and right edges, Cause unnecessary clipRect and affect performance 

Incredibly this bug has existed for several years

## Related Issues

Bug found during code review
Fixes https://github.com/flutter/flutter/issues/61434

## Tests

I added the following tests:

'SingleChildScrollView does not clipRect when the child size equals parent size'

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/widgets/single_child_scroll_view.dart [hunk 1]
```diff
@@ -568,7 +568,14 @@ class _RenderSingleChildViewport extends RenderBox with RenderObjectWithChildMix
 
   bool _shouldClipAtPaintOffset(Offset paintOffset) {
     assert(child != null);
-    return paintOffset < Offset.zero || !(Offset.zero & size).contains((paintOffset & child.size).bottomRight);
+    if (paintOffset.dx < 0 || paintOffset.dy < 0)
+      return true;
+
+    final Rect childRect = paintOffset & child.size;
+    if (childRect.left > size.width || childRect.bottom > size.height)
```
[the remark this answers is not in the document]
Comment (xu-baolin) [reply packages/flutter/lib/src/widgets/single_child_scroll_view.dart]: fixed, NICE code review, and clean code.

[the remark this answers is not in the document]
Comment (xu-baolin) [reply packages/flutter/lib/src/widgets/single_child_scroll_view.dart]: I will add a unit test too, It seems too easy to make mistakes here ^.^

[the remark this answers is not in the document]
Comment (xu-baolin) [reply packages/flutter/lib/src/widgets/single_child_scroll_view.dart]: Done.

**Review by liyuqian — APPROVED** [review]
LGTM

**Comment by liyuqian** [comment]
This is breaking some Google internal golden images as noted in b/163420457. @goderbauer @Piinks @renyou: Should we do the following, or directly modify the Google internal tests? If we need to add a flag, I wonder what the best way is to expose that flag. Adding it as an optional named parameter to `SingleChildScrollView` constructor with `deprecation` warning and remove it as soon as possible?

(1) revert the commit,
(2) submit a new commit with a flag to preserve the old behavior by default, but allow new behavior to be opted in
(3) enable the flag in Google internal tests
(4) remove the flag in Flutter

**Comment by xu-baolin** [comment]
@liyuqian Hi, I want to ask a question, how can I capture a special frame SkPicture by `flutter screenshot --type=skia --observatory-uri`, I found it only captures the 1st frame of the APP.

**Comment by liyuqian** [comment]
It should capture the current frame of the app instead of the 1st frame of
the app. In what version of Flutter did you find this problem? I'm
currently having difficulty in visualizing the skp due to
https://bugs.chromium.org/p/skia/issues/detail?id=10588

**Comment by liyuqian** [comment]
Just chatted with @renyou offline. He'll check if the failures are just small AA changes. If so, we can just rebaseline the images without having to do anything extra regarding this commit.

**Pull request merged**